server/src/main/java/org/elasticsearch/cluster/coordination/ElectionSchedulerFactory.java

```diff
@@ -0,0 +1,194 @@
+
+    public static final Setting<TimeValue> ELECTION_MAX_TIMEOUT_SETTING = Setting.timeSetting(ELECTION_MAX_TIMEOUT_SETTING_KEY,
+        TimeValue.timeValueSeconds(10), TimeValue.timeValueMillis(200), TimeValue.timeValueSeconds(300), Property.NodeScope);
+
+    private final TimeValue initialTimeout;
+    private final TimeValue backoffTime;
+    private final TimeValue maxTimeout;
+    private final ThreadPool threadPool;
+    private final Random random;
+
+    public ElectionSchedulerFactory(Settings settings, Random random, ThreadPool threadPool) {
+        super(settings);
+
+        this.random = random;
+        this.threadPool = threadPool;
+
+        initialTimeout = ELECTION_INITIAL_TIMEOUT_SETTING.get(settings);
+        backoffTime = ELECTION_BACK_OFF_TIME_SETTING.get(settings);
+        maxTimeout = ELECTION_MAX_TIMEOUT_SETTING.get(settings);
+    }
+
+    /**
+     * Start the process to schedule repeated election attempts.
+     *
+     * @param gracePeriod       An initial period to wait before attempting the first election.
+     * @param scheduledRunnable The action to run each time an election should be attempted.
+     */
+    public Releasable startElectionScheduler(TimeValue gracePeriod, Runnable scheduledRunnable) {
+        final ElectionScheduler scheduler = new ElectionScheduler();
+        scheduler.scheduleNextElection(gracePeriod, scheduledRunnable);
+        return scheduler;
+    }
+
+    @SuppressForbidden(reason = "Argument to Math.abs() is definitely not Long.MIN_VALUE")
+    private static long nonNegative(long n) {
+        return n == Long.MIN_VALUE ? 0 : Math.abs(n);
+    }
+
+    /**
+     * @param randomNumber a randomly-chosen long
+     * @param upperBound   inclusive upper bound
+     * @return a number in the range (0, upperBound]
+     */
+    // package-private for testing
+    static long toPositiveLongAtMost(long randomNumber, long upperBound) {
+        assert 0 < upperBound : upperBound;
+        return nonNegative(randomNumber) % upperBound + 1;
+    }
+
+    @Override
+    public String toString() {
+        return "ElectionSchedulerFactory{" +
+            "initialTimeout=" + initialTimeout +
+            ", backoffTime=" + backoffTime +
+            ", maxTimeout=" + maxTimeout +
+            '}';
+    }
+
+    private class ElectionScheduler implements Releasable {
+        private final AtomicBoolean isClosed = new AtomicBoolean();
+        private final AtomicLong attempt = new AtomicLong();
+
+        void scheduleNextElection(final TimeValue gracePeriod, final Runnable scheduledRunnable) {
+            if (isClosed.get()) {
+                logger.debug("{} not scheduling election", this);
+                return;
+            }
+
+            final long thisAttempt = attempt.getAndIncrement();
+            final long maxDelayMillis = Math.min(maxTimeout.millis(), initialTimeout.millis() + thisAttempt * backoffTime.millis());
```

Comment:
You mentioned the overflow that could happen here. Can you please add a safeguard? Otherwise I might not be able to sleep anymore 😨

Reply:
Ok. It's a bit daft because now that there are upper limits on the settings it would take solidly more than a million years to overflow, but there you go.

Reply:
if you put it like that, you can undo c08cd8b